src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const result = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        newState[key] = action.extraData[key];
+      }
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      newState = {};
```

Comment:
The 'clear' case is correctly resetting the state to an empty object. However, since this action is independent of the current state, you could optimize by not copying the currentState before this action.